README.md
# Dragon Warrior Randomizer #

This application is a randomizer for Dragon Warrior for NES. This randomizer
will work on either version of the ROM.

If you enjoy Dragon Warrior Randomizer, come join our
[discord community](https://discord.gg/SmHeCkD)!

## The randomizer has the following features: ##

#### Random overworld map generation ####

A new overworld map is generated with each seed. Checks are used to make
sure all towns and caves are accessible. On generated maps, the Rainbow Drop
is used just outside Charlock to create a bridge over the moat.

#### Randomizing chest contents ####

Chest contents throughout the game are randomized. Checks are used to ensure that
quest items (Staff of Rain, Silver Harp, Stones of Sunlight) are not located in 
Charlock castle. Erdrick's Tablet is removed.

#### Randomizing enemy attack patterns ####

This will give some enemies spells they did not originally have. This means you
could have a Red Slime with SLEEP, a Wizard with Fire Breath, or any other 
combination!

#### Randomizing enemy zones ####

The locations of enemies will be changed. Randomizer ensures that enemies near
Tantegel Castle are not too tough to defeat at low levels, but beyond that,
anything goes.

#### Random enemy stats, XP, and gold ####

Enemy statistics (strength, agility, HP) can be randomized, as can the amount of
gold and experience you receive from them.

#### Boosting XP/gold drops ####

Enemies will now give you more XP and Gold, in line with the SNES remake of the
game. XP and gold drops are limited to 255 due to game engine limitations.

#### Lowering MP requirements ####

MP requirements for spells are adjusted to be the same as those in the remake.

#### Randomized weapon shops ####

Weapon shop inventory will now be random. Every shop will have 5-6 random items.

#### Random XP requirements ####

Experience requirements to obtain a new level can be randomized.

#### Random weapon prices ####

Weapon prices can be randomized. This does not affect their capability.

#### Shuffled "searched for" items ####

The Fairy Flute, Erdrick's Armor, and Erdrick's Token will now be shuffled, so
for example the token could be in Kol instead of the flute, and the armor on the
overworld. Erdrick's Token or the Flute could also be in a random chest instead
of on the ground. In addition, the item's location on the overworld is random. The
NPC in Cantlin can tell you where it is.

#### Randomized player attributes ####

Your character's Strength, Agility, HP and MP are randomly generated.

#### Randomized spell learning ####

Spells learning is now randomized. Any spell can be learned anywhere from level 1 to
level 16.

#### Buffed HEAL and HURT ####

The HEAL spell has a range from 17-25. The HURT spell has a range from 9-16.

#### Faster leveling ####

The number of experience points required for level ups can be adjusted with the
fast leveling and very fast leveling options. These require 75% and 50% of the 
normal requirements for the game, respectively.

#### Cursed Princess mode ####

Get Princess Gwaelin to take a cursed belt from your inventory when you bring
her to the king to win the game. The game can also still be completed in the
normal manner.

#### Three's Company mode ####

Bring Princess Gwaelin to the Dragonlord and agree to join him in ruling the
world to win in this mode. The game can also still be completed in the normal
manner.

#### Alternate player sprites ####

Over 60 alternate sprites to choose from for the main character.

#### Other changes ####
The Death Necklace is always available in a chest. Putting it on gives you 
+10 attack power at the cost of 25% of your max HP and being cursed. The
Fighter's Ring works (+2 ATK). Other options include faster text, open Charlock
(no Rainbow Drop required), shortened Charlock dungeon, disabling magic keys, 
or more swamp on the overworld.

## Running the randomizer ##

##### Windows #####

Simply click on the executable in your file manager. On Windows, you may want to
place this somewhere easy to find, such as on the Desktop.

##### Mac OSX #####

On Mac OSX, DWRandomizer will require the installation of Qt5. I recommend
installing this via [homebrew](https://brew.sh):
```
brew install qt5
```

The *.app* folder contained in the zip file may be placed inside the
Applications folder for easy access from Launchpad.

##### Linux #####

Qt5 is also required on Linux. This should be available in your package manager.
If you'd prefer not to install this, there is a command line version also
available in the download which does not require Qt.

### Usage ###

Once the application is running, simply select your Dragon Warrior ROM file from
The `ROM File` box, and choose an output directory. The `Seed` box will be pre-
filled with a random seed number. You can generate a new one with the `Random`
button, or choose your own by entering it in the box. The `Flags` box is where
you enter flags for the options you would like to use, or simply choose them with
the check boxes below. If you're not sure, click `Defaults` for the options most
players use. Next, choose your player sprite. `Loto` is the default original 
sprite. There are other sprites to choose from, including sprites from other
famous games. Now click `Randomize`, and you should get a message saying your
ROM has been created. It will be in the directory you specified. Simply run this
using your favorite emulator or flash cart.

##### Options #####

* Shuffle Chests & Search Items: Self-explanatory. Changes what are in chests and search locations.
* Random Weapon Shops: Randomizes what items are available in all weapon shops.
* Random Growth: Randomize the stat changes you receive on each level up.
* Random Spell Learning: Randomizes the level at which you learn each spell.
* Random Enemy Attacks: Enemy spells and abilities are randomized.
  * Dragonlord form 2 may also have HEAL or SLEEP.
* Random Enemy Zones: Enemy locations are random.
* Random Enemy XP & Gold: Randomize the amount of XP and Gold received from enemies
* Random Weapon Prices: Weapon & armor prices are randomized.
* Random XP Requirements: Experience requirements for each level are randomized.
* Enable Menu Wrapping: Enable wrapping of the cursor from top to bottom in menus.
* Enable Death Necklace: Adds additional functionality to the Death Necklace. +10 ATK and -25% HP.
* Shuffle Music: Shuffles which tunes you hear in each location. Battle music is unchanged.
* Disable Music: Disables most game music. Battle music is not disabled, and music plays after an inn stay. These are known issues.
* Big Swamp: Makes approximately 75% of the overworld contain swamp tiles. Ouch!
* Fast Text: Speeds up all text in dialogs.
* Speed Hacks: Speeds up other aspects of the game (spells, fairy flute, death music, etc.)
* Open Charlock: No need to go around collecting items to make a bridge. Just walk right into Charlock!
* Short Charlock: Dragonlord skimped on the dungeon so you go straight to his throne room on entering Charlock.
* Don't Require Magic Keys: Doors will open without a key. The key shop owners have gone on strike.
* Cursed Princess: Make the princess take a cursed belt to win.
* Three's Company: Bring the princess to the Dragonlord and join him to win.

## Compiling from source ##
If you'd prefer to compile your own version from source code, here are some
instructions.

##### Windows #####

The easiest way I've found *so far* to build dwrandomizer on Windows is using
Ubuntu in Windows Subsystem for Linux on Windows 10 along with the MXE
cross-compiler. Setting up this environment is rather involved, but I'll try
to cover it briefly.

First, set up [Windows Subsystem for Linux](https://docs.microsoft.com/en-us/windows/wsl/install-win10)

Next, download [mxe](https://github.com/mxe/mxe) into your Linux filesystem.
Install the needed dependencies and build it with qt5 support. This may take a
while to compile. 

```
apt update
apt -y install automake autopoint bison flex gperf intltool libgdk-pixbuf2.0-dev libtool-bin lzip make p7zip-full ruby unzip
make qt5
```

After this completes, navigate to the `dwrandomizer` source directory in your
terminal and run the following (replacing `/path/to/mxe` with the path to the
mxe directory you just created):

```
mkdir build
cd build
/path/to/mxe/usr/x86_64-w64-mingw32.static/qt5/bin/qmake ../dwrandomizer.pro
make
```

This will compile dwrandomizer and the new executable should be in the release
directory.

##### Mac OSX #####

First, install Qt5 using homebrew.
```
brew install qt5
```
Now, download the randomizer source code and change to that directory.
```
mkdir build
cd build
/usr/local/Cellar/qt5/5.9.1/bin/qmake ../dwrandomizer.pro
make
```
The Qt version number may be different. This will create an app folder that you
can run from finder, or from the command line simply run
```
dwrandomizer.app/Contents/MacOS/dwrandomizer
```

##### Linux #####

First, install Qt5 using your package manager. Download the randomizer source
code, then change to the directory containing the source code.

```
mkdir build
cd build
qmake ../dwrandomizer.pro
make
```

Some compiling will happen, and that's it! you should have a working binary in 
the current directory.

##### Emscripten #####

Install emscripten from https://emscripten.org/docs/getting_started/downloads.html

Instructions for MacOS and linux:
```bash
mkdir build
cd build
cmake -DCMAKE_TOOLCHAIN_FILE=<path to emscripten>/libexec/cmake/Modules/Platform/Emscripten.cmake ..
make dwrandomizer-cli
```

You'll then have `dwrandomizer-cli.wasm` and `dwrandomizer-cli.js` files, which can be served along with the files in the
 `emscripten` directory. For example, continuing from the same `build` directory as above:
```bash
cp ../emscripten/* .
python -m http.server
```
Then open a web browser to http://localhost:8000


## FAQ ##

#### What is the "checksum" when generating a ROM? ####

The program outputs a checksum when generating a new ROM. The checksum
can be used to verify that 2 instances of the game are the same, e.g. the
input ROM is the same and the same seed and flags were used. This is 
mostly useful for verifying ROMs in races, where normally all players play the 
same randomized game.


CHANGELOG.md
# 3.0 - So many changes! #
###### Released Nov 30, 2021 ######
* Entirely new web-based user interface
* Added tri-state flags for most options. This allows an option to randomly be on or off.
  - Two flags are exceptions, which are "partial" flags
    - "Summer Sale" will only discount about half of the items when set to "partial"
    - "Treasure Guards" will set only about half of important items to be guarded.
* Added tracking of certain stats which will be displayed after the game ending. The following are tracked:
  - Encounters, defeats, and deaths for each monster.
  - The number of times you used each spell.
  - Deaths - The number of times you died.
  - Attacks - The number of times you used the "Attack" command.
  - Crits - The number of times you had an "Excellent Move".
  - Dodges - The number of times an enemy dodged
  - Blocks - The number of times you were prevented from running.
  - Ambushes - The number of times a monster attacked you first.
  - Bonks - The number of times you ran into an obstacle.
  - Play time - In game play time from when you selected message speed until the "Until the dragons return again" text displays.
    - This includes time on the title screen if you have started the game and then reset.
* Added a flag for enabling random chest placement.
* Added stair shuffle to randomize where dungeon stairs go.
* Added alternate algorithm for determining run chances.
  - With this flag on, chances of running will be determines by which map you are on rather than the monster difficulty.
  - This is the same algorithm used when "Random Monster Stats" is on.
* Added "Summer Sale" flag which will discount all weapons and armor by 35-65%
* Added "Random Map Size" option to allow generation of smaller maps.
* Added "Treasure Guards" where a mid-level spike encounter will guard important items.
  - Guarded items are Erdrick's Armor & Sword, Silver Harp, Erdrick's Token, and Stones of Sunlight.
* Added "Randomly Rotate/Mirror Dungeons" which will change the orientation of dungeon maps
  - Maps can be rotated 90, 180, or 270 degrees in addition to being vertically or horizontally mirrored.
* Added a summary tab which will display a summary of what options are set to.
* Added sorting of player inventory based on item index
* Added display of what spells you learned at level up (instead of "thou hast learned a spell").
* Added support for 2-byte XP and Gold drops from monsters
  - Red Dragons now give 350 XP.
  - Golems now give 350 XP.
  - Goldmen now give 650 GOLD.
* Adjusted the player strength growth curve.
* Adjusted the curves for random XP and Gold drops for monsters based on the new values.
* "Short Charlock" now includes the top floor of Charlock unless "Stair Shuffle" is also on.
* Added new sprites:
  - FF4 Edge
  - Plumplim
  - DoaE Guan Yu
  - DoaE Zhang Fei
  - DoaE Zhao Yun
  - DoaE Huang Zhong
  - DoaE Ma Chao
  - DoaE Zhuge Liang
  - DoaE Jiang Wei
  - DoaE Lu Bu
  - Bo Jackson
  - Jerry Rice
  - Barry Sanders

# 2.2.1 - Some balance adjustments
##### Released July 06, 2021 #####
* Maximum STOPSPELL resist is now 9/16
* Maximum SLEEP/DL2 breath chance is now 50%
* New Flag: Scaled Metal Slime XP (32 times level)
  - With this flag Metal Slimes will give XP equal to 32 times your level or 255, whichever is less.

# 2.2 - Armor in Charlock? Yes. And much more!
##### Released Mar 01, 2021 #####
* Added the ability for Erdrick's Armor to be in a chest
* The 3 search spots can now have any unique item: Sword, Armor, Token, Harp, Stones, Death Necklace, Fairy Flute.
   - Death Necklace on a search spot also creates a mini gold grind there.
* Added the ability to use torches and fairy water in battle
* Added a flag for forcing heal/hurt before healmore/hurtmore
* Added a "No Hurtmore" flag
* Changed "Vanilla Map" to "Random Map"
  - Opposite logic, of course
* Fixed picking up the token again after getting the Rainbow Drop
  - The chest/search spot will now be empty instead if you have the drop
* Moved Challenge mode sprites to be flags on the new Challenge tab
* Removed the extra town/cave locations from vanilla map
* Major overhaul of all stat generation
  - Added "make stats consistent" option for this mode for those who want the stats to be a little more sane.
* Also modified algorithms for random weapon prices and random XP.
* Modified the harp to pull lower-tier monsters (Zone 1&2)
* Rimuldar will no longer be required on the first continent if there is an accessible key in a cave
* Major speed increase to map generation (thanks dvj!)
* Removed stray sheild sprite in "Invisible hero" (thanks dwedit!)
* Added an option to disable screen flashing for spell casts
* Added "Permanent Torch" option
  - This will make the minimum viewable area 3x3 in dark areas
  - Torches will have no effect, but radiant still works as expected
* Fixed a bug where HP was not boosted when "Randomize Growth" was off in big swamp mode.
* Made sure there is no item on the overworld if "No Numbers" is enabled.

* Fixed an issue where towns were occasionally placed in improper spots on non-random maps 
* Fixed random sprite choice to not be deterministic
  - The random sprite chosen will no longer depend on the seed number
* Added new sprites:
  - Ms. Pacman Sue
  - FF4 Cecil
  - Yoshi
  - Fixed an error in the Kirby princess carrying sprite

# 2.1.2 - Bug fixes #
##### Released July 28, 2019 #####
* Fixed the slime and red slime sprites

# 2.1.1 - Bug fixes #
##### Released July 28, 2019 #####

* Fixed a bug where selecting the default hero sprite would crash the randomizer
* Fixed a bounds check error in the map code which could crash the randomizer
* Some minor edits to a few sprites.

# 2.1 - More ways to WIN! #
##### Released July 27, 2019 #####

* Cursed Princess mode!
  - In this mode, you can win the game by getting Gwaelin to take a cursed belt from your inventory when you return her to the king.
* Three's Company mode!

hud.lua

debug = {
	hud = true,
  cheats_enabled = false,
  cheats = function()
    player:set_experience(65535)
    player:set_gold(65535)
    memory.writebyte(0xbe, 255) -- best equipment
    memory.writebyte(0xbf, 6)   -- 6 herbs
    memory.writebyte(0xc0, 6)   -- 6 keys
    memory.writebyte(0xc1, 14)  -- rainbow drop
  end 
}

features = {
  herb_store = true, -- buy herbs anywhere
  enemy_hp = true,  -- show enemy hit points
  repulsive = false, -- no lower level enemy encounters
}

levels = {}

TILE = {
  GRASS = 0,
  DESERT = 1,
  HILLS = 2,
  STAIRS_UP = 3,
  TILES = 4,
  STAIRS_DOWN = 5,
  SWAMP = 6,
  TOWN = 7,
  CAVE = 8,
  CASTLE = 10,
  BRIDGE = 11,
  FOREST = 12,
  CHEST = 13,
  FORCE_FIELD = 14
}

tile_names = {
  "Grass",
  "Desert",
  "Hill",
  "Stairs Up",
  "Brick",
  "Stairs Down",
  "Swamp",
  "Town",
  "Cave",
  "Castle",
  "Bridge",
  "Forest",
  "Chest",
  "Barrier",
}

maps = {
  "Alefgard",
  "Charlock",
  "Hauksness",
  "Tantegel",
  "Tantegel Throne Room",
  "Charlock Throne Room",
  "Kol",
  "Brecconary",
  "Garinham",
  "Cantlin",
  "Rimuldar",
  "Tantegel Basement",
  "Northern Shrine",
  "Southern Shrine",
  "Charlock Cave Lv 1",
  "Charlock Cave Lv 2",
  "Charlock Cave Lv 3",
  "Charlock Cave Lv 4",
  "Charlock Cave Lv 5",
  "Charlock Cave Lv 6",
  "Swamp Cave",
  "Mountain Cave",
  "Mountain Cave Lv 2",
  "Garin's Grave Lv 1",
  "Garin's Grave Lv 2",
  "Garin's Grave Lv 3",
  "Garin's Grave Lv 4",
  "Erdrick's Cave",
  "Erdrick's Cave Lv 2"
}

strings = {
  encounter = {
    "A %s draws near!",
    "Hark! A %s approacheth!",
    "What is that I see? Why it's a %s!",
    "I have no quarrel with you Sir %s, but I must pass.",
    "You got something for me %s? Bring it!"
  },
  enemyrun = {
    "I suppose he didst not likest me much.",
    "Comest back here thee coward!",
    "Oh running away eh? Come back here and fight me! I'll bite your legs off!",
    "Oh comest on! Why did thee even comest over if thee were just to runnest away?",
    "It was in quite a hurry.",
    "That's right! You get on down the road and don't come back!",
    "Fine, just goest. Leavest me liketh everyone else. Well, excepteth that clingy princess.",
    "Wait what just happened? That %s just disappeared!"
  },
  playerrun = {
    "Whew! That was scary!",
    "I have made my escape from the vile %s",
    "I just felt like running. So I did.",
    "You can't catch me son! Don't even try!",
    "It was a fierce beast! With big fangs! What? Didn't you see the bones?",
    "I think I soiled my armor.",
    "Who, me? I wasn't running from it, I just had somewhere to be...that wasn't anywhere near that %s."
  },
  enemydefeat = {
    "I hath done well in defeating the %s.",
    "Oh! You got wrecketh bro!",
    "Thou just stayeth down there chump!",
    "There ain't gonna be no rematch!",
    "Did you see that? There were like 20 of them! No match for me though.",
    "That was easy.",
    "Seriously? That's all you got? Where's the REAL challenger?",
    "You can't touch this.",
    "I attack the %s! It's super effective!",
    "Who has two thumbs and just put thee down? This guy!",
    "Who's the hero? Who is the hero? That's right, I'm the hero."

  },
  lowhp = {
    "Uuungh... I don't feel so well",
    "MEDIC!",
    "I'm kinda losing a lot of blood here...",
    "'Tis but a scratch!",
    "Oh, it's just a flesh wound.",
    "It is getting dark...I am not long for this world! Oh, nevermind, my helmet was on backward.",
    'So what are "hit points" anyway? More is better, right?'
  },
  playerdefeat = {
    "Ouch.",
    "That really hurt, Charlie!",
    "Oy! Gevalt!",
    "I'll be back.",
    "I'm not quite dead yet. Oh, nevermind, yes I am.",
    "Oh great king! A little help here please!",
    "Well that didn't go quite like I envisioned it.",
    "Thou are at fault for this! Thou! Thou! Not I! Thou art a terrible tactician!",
    "They telleth me to goest toward the light..."
  }
}



function say(str) 
  print(string.format("%s\n", str))
end

function battle_message(strings, enemy_type)
--  if enemy_type == nil then
--    say(strings[math.random(#strings)])
--  else
--    local enemy_name = enemy.types[enemy_type]
--    if enemy_name ~= nil then
--      say(strings[math.random(#strings)]:format(enemy_name))
--    end
--  end
end


in_battle = false
pre_battle = false
function battle_mode (battling, reset_enemy_hp)
  if battling ~= nil then
    in_battle = battling
    pre_battle = battling
    player.valid_tile = false
    if not battling then
      if (reset_enemy_hp == true) then
        -- overwrite monster hp to avoid confusion
        enemy:set_hp(0)
      end
    end
  end
  return in_battle
end

Player = {
  level = 0,
  hp = 0,
  mp = 0,
  gold = 0,
  experience = 0,
  change = {
    level = 0,
    hp = 0,
    mp = 0,
    gold = 0,
    experience = 0
  },
  herbs = 0,
  keys = 0,
  map_x = 0,
  map_y = 0,
--   player_x = 0,
--   player_y = 0,
--   current_map = 0,
  tile = 0,
  last_tile = 0,
  valid_tile = false,

}

function Player.update (self)

  --level cap
  if features.level_cap ~= nil then
    local exp_cap = self.levels[features.level_cap]
    if self:get_experience() > exp_cap then
      self:set_experience(exp_cap)
    end
  end

  -- read in the values from memory.
  local level = self:get_level()
  local hp = self:get_hp()
  local mp = self:get_mp()
  local gold = self:get_gold()
  local experience = self:get_experience()

  -- update the changes.
  self.change.level = level - self.level
  self.change.hp = hp - self.hp
  self.change.mp = mp - self.mp
  self.change.gold = gold - self.gold
  self.change.experience = experience - self.experience

  -- update the object variables.
  self.level = level
  self.hp = hp
  self.mp = mp
  self.gold = gold
  self.experience = experience

  self.herbs = self.get_herbs()
  self.keys = self:get_keys()

  self.last_tile = self.tile
  self.tile = self:get_tile()
  local map_x = self:get_x()
  local map_y = self.get_y()
  if (map_x ~= self.map_x or map_y ~= self.map_y) then
    if in_battle then
      battle_mode(false, true)
    end
    self.valid_tile = true
  end
  if (self.current_map ~= self.get_map()) then
    self.valid_tile = false
  end
  self.map_x = map_x
  self.map_y = map_y
--   self.player_x = memory.readbyte(0x3a)
--   self.player_y = memory.readbyte(0x3b)
   self.current_map = self:get_map()

  if hp == 0 and self.change.hp ~= 0 and player:get_map() > 0 then
    battle_message(strings.playerdefeat, player:get_tile()+1)
  end

end

function Player.get_tile(self)
  return memory.readbyte(0xe0)
end

function Player.get_x(self)
  return memory.readbyte(0x8e)
end

function Player.get_y(self)
  return memory.readbyte(0x8f)
end

function Player.get_herbs(self)
  return memory.readbyte(0xc0)
end

function Player.get_keys(self)
  return memory.readbyte(0xbf)
end

function Player.get_level(self)
  return memory.readbyte(0xc7)
end

function Player.next_level(self)
  if self:get_level() == features.level_cap then
    return 0
  end
  for i=1,#self.levels do
    if self.levels[i] > self:get_experience() then
      return self.levels[i] - self:get_experience()
    end
  end
  return 0
end

function Player.get_hp(self)
  return memory.readbyte(0xc5)
end

function Player.max_hp(self)
  return memory.readbyte(0xca)
end

function Player.get_mp(self)
  return memory.readbyte(0xc6)
end

function Player.add_hp (self, amount)
  return self:set_hp(self.hp + amount)
end

function Player.set_hp (self, amount)
  if (amount > 255 or amount < 0) then
    return false
  end
  self.hp = amount
  memory.writebyte(0xc5, amount)
  return true
end

function Player.max_mp(self)
  return memory.readbyte(0xcb)
end

function Player.get_map(self)
  return memory.readbyte(0x45)
end

function Player.add_mp (self, amount)
  return self:set_mp(self.mp + amount)
end

function Player.set_mp (self, amount)
  if (amount > 255 or amount < 0) then
    return false
  end
  self.mp = amount
  memory.writebyte(0xc6, amount)
  return true
end

function Player.get_gold(self)
  return memory.readbyte(0xbd) * 256 + memory.readbyte(0xbc)
end

function Player.add_gold (self, amount)
  return self:set_gold(self.gold + amount)
end

function Player.set_gold (self, amount)
  if (amount > 65535 or amount < 0) then
    return false
  end
  self.gold = amount
  memory.writebyte(0xbd, amount / 256)
  memory.writebyte(0xbc, amount % 256)
  return true
end

function Player.get_experience(self)
  return memory.readbyte(0xbb) * 256 + memory.readbyte(0xba)
end

function Player.add_experience (self, amount)
  return self:set_experience(self.experience + amount)
end

function Player.set_experience (self, amount)
  if (amount > 65535 or amount < 0) then
    return false
  end
  self.experience = amount
  memory.writebyte(0xbb, self.experience / 256)
  memory.writebyte(0xba, self.experience % 256)
end


Enemy = {
  hp = 0,
  change = {
    hp = 0
  },
  types = {
    "Slime",  -- 0
    "Red Slime",
    "Drakee",
    "Ghost",
    "Magician",
    "Magidrakee", -- 5
    "Scorpion",
    "Druin",
    "Poltergeist",
    "Droll",
    "Drakeema",  --10
    "Skeleton",
    "Warlock",
    "Metal Scorpion",
    "Wolf",
    "Wraith",  --15
    "Metal Slime",
    "Specter",
    "Wolflord",
    "Druinlord",
    "Drollmagi",  --20
    "Wyvern",
    "Rogue Scorpion",
    "Wraith Knight",
    "Golem",
    "Goldman",  -- 25
    "Knight",
    "Magiwyvern",
    "Demon Knight",
    "Werewolf",
    "Green Dragon",  -- 30
    "Starwyvern",
    "Wizard",
    "Axe Knight",
    "Blue Dragon",
    "Stoneman", --35
    "Armored Knight",
    "Red Dragon",
    "Dragonlord",  --first form
    "Dragonlord"  --second form

  }
}

function Enemy.update (self)
  -- read in the values from memory.
  hp = memory.readbyte(0xe2)
  -- update the changes.
  self.change.hp = hp - self.hp
  -- update the object variables.
  self.hp = hp

  -- update battle status
  if not in_battle and self.change.hp ~= 0 then
    battle_mode(true, false)
  end

  -- hit points wrap below zero, so check for large increases.
  if in_battle and (self.hp == 0 or self.change.hp > 160) then
    battle_mode(false, false)
    battle_message(strings.enemydefeat, player:get_tile()+1)
  end

end

function Enemy.show_hp (self)
  if (in_battle and features.enemy_hp) then 
    gui.drawbox(152, 134, 190, 144, "black")
    gui.text(154, 136, string.format( "HP %3d", self.hp), "white", "black")
  end
end

function Enemy.set_hp(self, hp)
  if (hp > 255 or hp < 0) then
    return false
  end
  self.hp = hp
  memory.writebyte(0xe2, hp)
  return true
end

function Enemy.hud (self)
  enemy_index = memory.readbyte(0xe0)
  enemy_type = enemy.types[enemy_index+1]
  lbound = 186
  if enemy_type ~= nil then
    gui.drawbox(lbound, 64, 260, 144, "black")
    gui.text(lbound, 64, enemy_type, "white", "black")
    gui.text(lbound, 74, string.format( "HP %3d", self.hp), "white", "black")
    gui.text(lbound, 84, 
      string.format("Str %3d", rom.readbyte(0x5e5b + enemy_index * 16))
      , "white", "black")
    gui.text(lbound, 94, 
      string.format("Agi %3d", rom.readbyte(0x5e5c + enemy_index * 16))
      , "white", "black")
    -- show the special attacks
    specialset1 = { "Sleep", "Stopspell", "Heal", "Healmore" }
    specialset2 = { "Hurt", "Hurtmore", "fire", "fire 2" }
    specials = rom.readbyte(0x5e5e + enemy_index * 16)
    ypos = 104
    spectype = specialset1[math.floor(specials / 64) + 1]
    specchance = math.floor(specials / 16 % 4) * 25
    if specchance > 0 then
      gui.text(lbound, ypos, string.format("%s (%2d%%)", spectype, specchance), 
        "white", "black")
      ypos = ypos + 10
    end
    spectype = specialset2[math.floor((specials / 4) % 4) + 1]
    specchance = (specials % 4) * 25
    if specchance > 0 then
      gui.text(lbound, ypos, string.format("%s (%2d%%)", spectype, specchance), 
        "white", "black")
    end
  end
end
-- 
--  Draws any hud elements, such as the enemy hit points
--  and debug info
-- 
function overlay()
  if debug.hud then
    current_map = maps[memory.readbyte(0x45)]
    if current_map ~= nil then
      gui.drawbox(0, 0, 255, 16, "black")
      gui.text(8, 8, 
        string.format("%s (%3d,%3d)", current_map, memory.readbyte(0x8e), 
          memory.readbyte(0x8f)), 
        "white", "black")
      if in_battle then
        enemy:hud()
      end
      if memory.readbyte(0x45) == 1 then
        zone = math.floor(player.map_x / 15) + math.floor(player.map_y / 15) * 8
        if zone % 2 == 1 then
          zone = rom.readbyte(zone/2 + 0xf532)
          zone = math.floor(zone % 16)
        else
         zone = rom.readbyte(zone/2 + 0xf532)
         zone = math.floor(zone / 16)
        end
        gui.text(215, 8,
            string.format( "Zone %d", zone), "white", "black")
      end
    end
  else
    enemy:show_hp()
  end
end

function read_levels()
  for i=1,31 do
    levels[i] = rom.readbyte(0xf36c+(i*2)) * 256 + rom.readbyte(0xf36b+(i*2))
  end
end

function update()
  if (player:get_map() > 0) then
    -- update the player and enemy info every 1/4 second
    if (emu.framecount() % 15 == 0) then
      player:update()
      enemy:update()
    end

    if features.repulsive then
      memory.writebyte(0xdb, 0xff)
    end
  end
end


-- for battle detection (enemies or you running away)
function running(address)
  if (player:get_map() > 0) then
    if address == 0xefc8 then
      battle_message(strings.enemyrun, player:get_tile()+1)
    else
      battle_message(strings.playerrun, player:get_tile()+1)
    end
    battle_mode(false, true)
  end
end

-- A thing draws near!
function encounter(address)
  if (player:get_map() > 0) then
    player.valid_tile = false
    battle_message(strings.encounter, memory.readbyte(0x3c)+1)
    pre_battle = true
  end
end

function cheats()
  if debug.cheats_enabled and debug.cheats then
    if (memory.readbyte(0x45) == 1) then
      debug.cheats()
      debug.cheats = false
    end
  end
end


-- main loop
function main()
  -- savestate.load(savestate.object(1))
  memory.registerexecute(0xefc8, running)
  memory.registerexecute(0xe8a4, running)
  memory.registerexecute(0xcf44, encounter)
  memory.registerwrite(0x45, cheats)
  read_levels()

  player = Player
  enemy = Enemy
  player.last_command = emu.framecount()
  enemy:update()
  player:update()
  gui.register(overlay)
  emu.registerafter(update)

  while(true) do
    emu.frameadvance()
  end
end

main()
